subprojects/data-structures/src/main/java/org/gradle/internal/collect/PersistentList.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.collect;
+
+import org.gradle.internal.Cast;
+
+import javax.annotation.CheckReturnValue;
+import java.util.function.Consumer;
+
+public interface PersistentList<T> {
+    static <T> PersistentList<T> of() {
+        return Cast.uncheckedCast(Nil.NIL);
+    }
+
+    static <T> PersistentList<T> of(T first) {
+        return PersistentList.<T>of().plus(first);
+    }
```

Comment:
Can we have a method that takes `T...` as well?

Reply:
I added that method.